### PR TITLE
adding the -s flag to show_source

### LINKED ISSUE #718: Support `-s` option in `edit` and `show_source` commands
## Description

In `Pry`, users can use `show-source Bar#foo -s` to access `Foo#foo`'s source of the example below:

```rb
class Foo
  def foo
  end
end

class Bar < Foo
  def foo
    super
  end
end
```

This could be extremely helpful when debugging at times and it'd be great to support it in IRB.

### PR DESCRIPTION
**Pull request by paulreece**
This enhancement allows a user to add the -s flag if they want to access a methods origin definition.  It allows for chaining of multiple esses to further go up the classes as needed.

With the following code:
```ruby
class Foo
  def foo
  end
end

class Bar < Foo
  def foo
    super
  end
end
```

If you call `show_source Bar#foo -s` it will return:
```ruby
From: test.rb:4

  def foo
  end

```

Closes #718.

### CODE REVIEW

**Comment by paulreece** [comment]
Sorry for all of the pushes! My test file had a namespace clash with the `Foo` class that was being used by the other tests.  I just changed it to `Baz` instead. :)

File: test/irb/test_cmd.rb [hunk 1]
```diff
@@ -469,6 +469,17 @@ def test_show_source_method
...
     end
 
+    def test_show_source_method_s
+      File.write("#{@tmpdir}/bazbar.rb", "class Baz\n  def foo\n  end\nend\n\nclass Bar < Baz\n  def foo\n    super\n  end\nend\n")
+      out, err = execute_lines(
+        "irb_load '#{@tmpdir}/bazbar.rb'\n",
+        "show_source Bar#foo -s",
+      )
+      assert_empty err
+      p out
```
Comment (st0012) [inline test/irb/test_cmd.rb:479]: I think this is a leftover from debugging?

File: test/irb/test_cmd.rb [hunk 2]
```diff
@@ -469,6 +469,17 @@ def test_show_source_method
       assert_match(%r[/irb\/init\.rb], out)
     end
 
+    def test_show_source_method_s
+      File.write("#{@tmpdir}/bazbar.rb", "class Baz\n  def foo\n  end\nend\n\nclass Bar < Baz\n  def foo\n    super\n  end\nend\n")
```
Comment (st0012) [inline test/irb/test_cmd.rb:473]: Could we use heredoc for this?

File: test/irb/test_cmd.rb [hunk 3]
```diff
@@ -469,6 +469,17 @@ def test_show_source_method
       assert_match(%r[/irb\/init\.rb], out)
     end
 
+    def test_show_source_method_s
```
Comment (st0012) [inline test/irb/test_cmd.rb:472]: Can we have more tests to cover:

- Multiple super level like `-sss`
- When the level of flags, like `-sss`, exceeds the actual super chain of the method
- When the super owner doesn't support `instance_method` method
- When `-s` is accidentally mixed with other characters, like `-sd`
    - I think in this case it should be seen as `-s`

**Review by st0012 — COMMENTED** [review]
Thank you for the PR! Just few comments about testing.

File: test/irb/test_cmd.rb (hunk 2 above)
Comment (paulreece) [reply test/irb/test_cmd.rb:473]: Yes, apologies!

File: test/irb/test_cmd.rb (hunk 1 above)
Comment (paulreece) [reply test/irb/test_cmd.rb:479]: Yes, apologies!

File: test/irb/test_cmd.rb (hunk 3 above)
Comment (paulreece) [reply test/irb/test_cmd.rb:472]: These are all great edge cases.  I'll get to work on implementing them today!

Comment (paulreece) [reply test/irb/test_cmd.rb:472]: Okay I added functionality and tests for all of these including `heredoc` for the code.  Let me know what you think!

File: test/irb/test_cmd.rb [hunk 4]
```diff
@@ -469,6 +469,116 @@ def test_show_source_method
...
+            super
+          end
+        end
+      RUBY
+      File.write("#{@tmpdir}/bazbar.rb", code)
+      out, err = execute_lines(
+        "irb_load '#{@tmpdir}/bazbar.rb'\n",
+        "show_source Bar#foo -s",
+      )
+      assert_match(%r[Error: Couldn't locate a definition for Bar#foo -s\n=> nil\n], out)
```
Comment (st0012) [inline test/irb/test_cmd.rb:536]: Is it possible to change the message to `Couldn't locate the super definition for Bar#foo`?

File: lib/irb/source_finder.rb [hunk 5]
```diff
@@ -18,6 +18,11 @@ def initialize(irb_context)
 
     def find_source(signature)
       context_binding = @irb_context.workspace.binding
+      if signature.include? " -s"
+        signature, esses = signature.split(" -")
+        esses.count("^s").zero? ? s_count = esses.size : s_count = 1
```
Comment (st0012) [inline lib/irb/source_finder.rb:23]:
```suggestion
        s_count = esses.count("^s").zero? ? esses.size : 1
```

File: lib/irb/source_finder.rb [hunk 6]
```diff
@@ -26,6 +31,13 @@ def find_source(signature)
       when /\A(?<owner>[A-Z]\w*(::[A-Z]\w*)*)#(?<method>[^ :.]+)\z/ # Class#method
         owner = eval(Regexp.last_match[:owner], context_binding)
         method = Regexp.last_match[:method]
+        if esses
```
Comment (st0012) [inline lib/irb/source_finder.rb:34]: That's use `s_count` for this condition?

File: lib/irb/source_finder.rb [hunk 7]
```diff
@@ -26,6 +31,13 @@ def find_source(signature)
...
+          super_owner = owner
+          s_count.times {|s| super_owner = super_owner.superclass}
```
Comment (tompng) [inline lib/irb/source_finder.rb:36]: if there are too many -ssss, `super_owner.superclass` raises ``undefined method `superclass' for nil:NilClass``

File: lib/irb/source_finder.rb [hunk 8]
```diff
@@ -26,6 +31,13 @@ def find_source(signature)
...
+          if super_owner.respond_to?(:instance_method)
```
Comment (tompng) [inline lib/irb/source_finder.rb:37]: `super_owner` is always `nil` or `Class` here.
The original intention of `if owner.respond_to?(:instance_method)` is to handle a case that owner is not a class.
```ruby
show_source RUBY_VERSION#method
```

How about moving `if esses` logic inside `if owner.respond_to?(:instance_method)` ?
```ruby
if owner.respond_to?(:instance_method)
  if esses
     ...
  end
  methods = owner.instance_methods + ...
end
```

File: lib/irb/source_finder.rb (hunk 5 above)
Comment (paulreece) [reply lib/irb/source_finder.rb:23]: Much clearer.

File: lib/irb/source_finder.rb (hunk 6 above)
Comment (paulreece) [reply lib/irb/source_finder.rb:34]: We can use it, it will probably be clearer that way.

File: lib/irb/source_finder.rb (hunk 7 above)
Comment (paulreece) [reply lib/irb/source_finder.rb:36]: I added this test in there already in the previous commit:
```ruby
    def test_show_source_method_exceeds_super_chain
      code = <<~RUBY
        class Baz
          def foo
          end
        end
      
        class Bar < Baz
          def foo
            super
          end
        end
      RUBY
      File.write("#{@tmpdir}/bazbar.rb", code)
      out, err = execute_lines(
        "irb_load '#{@tmpdir}/bazbar.rb'\n",
        "show_source Bar#foo -ss",
      )
      assert_match(%r[Error: Couldn't locate a definition for Bar#foo -ss\n=> nil\n], out)
    end
```
Does that cover the use case you're referring to? If not I'm happy to add another.

File: lib/irb/source_finder.rb (hunk 8 above)
Comment (paulreece) [reply lib/irb/source_finder.rb:37]: This is cool context you've given for the conditional.  Right now the `signature` getting passed to the method is a string including the `-s` . For example if I call `show_source Baz#foo -s` in irb.  This method will receive `"Baz#foo -s"` as the `signature` string and error out in the case statement.  That's why I have the parsing logic living at the top there.

Would you suggest that we change the `Regex` in the conditional for this method and do the parsing there? Then we could switch the order of the case statements and resolve it that way.

```ruby
case signature
      when /\A[A-Z]\w*(::[A-Z]\w*)*\z/ # Const::Name
        eval(signature, context_binding) # trigger autoload
        base = context_binding.receiver.yield_self { |r| r.is_a?(Module) ? r : Object }
        file, line = base.const_source_location(signature)
      when /\A(?<owner>[A-Z]\w*(::[A-Z]\w*)*)#(?<method>[^ :.]+)\z/ # Class#method
        owner = eval(Regexp.last_match[:owner], context_binding)
        method = Regexp.last_match[:method]
        if s_count
          super_owner = owner
          s_count.times {|s| super_owner = super_owner.superclass}
          if super_owner.respond_to?(:instance_method)
            owner = super_owner
          end
        end
        if owner.respond_to?(:instance_method)
          methods = owner.instance_methods + owner.private_instance_methods
          file, line = owner.instance_method(method).source_location if methods.include?(method.to_sym)
        end
```

Comment (tompng) [reply lib/irb/source_finder.rb:36]: I was thinking of something like this. Parsing logic should be in top. Not change need for regexp.
```ruby
if owner.respond_to?(:instance_method) && s_count # to ensure owner is a class or module.
  # Maybe we need to check if owner is not a module, or support module too.
  super_owner = owner
  s_count.times do |s|
    # When super_owner reaches BasicObject, its superclass is nil.
    # To avoid calling nil.superclass, add `if super_owner.superclass` or `break unless super_owner` or use `&.`
    super_owner = super_owner.superclass
  end
  # super_owner is class or nil. To distinguish class and nil, respond_to?(:instance_method) is not needed.
  # Perhaps we can just `return unless super_owner` and show `Couldn't locate a definition`
  if super_owner needed.
    owner = super_owner
  end
end
if owner.respond_to?(:instance_method) # This check can be combined with the above one
  methods = owner.instance_methods + owner.private_instance_methods
  file, line = owner.instance_method(method).source_location if methods.include?(method.to_sym)
end
```
After two `if owner.respond_to?(:instance_method)` is combined, it will be like this, which I wanted to say in the previous comment
```ruby
if owner.respond_to?(:instance_method)
  if s_count # it was `if esses` before
     ...
  end
  methods = owner.instance_methods + ...
end
```
Or maybe using early return is more simple
```ruby
return unless owner.respond_to?(:instance_method)
if s_count
  ...
end
methods = ...
```

Comment (paulreece) [reply lib/irb/source_finder.rb:36]: Ah I understand now.  Sorry I misunderstood your original comment but I appreciate the second one for all of the cool ideas you mentioned!

File: lib/irb/cmd/show_source.rb [hunk 9]
```diff
@@ -58,4 +58,4 @@ def bold(str)
       end
     end
   end
-end
+end
```
Comment (ima1zumi) [inline lib/irb/cmd/show_source.rb:61]: The new line at the end of the file has been removed. Could you add it?

File: test/irb/test_cmd.rb (hunk 4 above)
Comment (paulreece) [reply test/irb/test_cmd.rb:536]: Just added this.  I moved the parsing to `ShowSource#execute" to accamodate this.

**Comment by paulreece** [comment]
FYI: the `truffleruby` tests seem to be erroring out with some gem dependency errors

File: lib/irb/cmd/show_source.rb [hunk 10]
```diff
@@ -27,11 +27,18 @@ def execute(str = nil)
...
+          s_count = esses.count("^s").zero? ? esses.size : 1
+          source = SourceFinder.new(@irb_context).find_source([str, s_count])
+        else
+          source = SourceFinder.new(@irb_context).find_source(str)
+        end
 
         if source
           show_source(source)
+        elsif s_count
+          puts "Error: Couldn't locate a definition for #{str}"
```
Comment (st0012) [inline lib/irb/cmd/show_source.rb:41]: Should it be this instead?
```suggestion
          puts "Error: Couldn't locate a super definition for #{str}"
```

Comment (paulreece) [reply lib/irb/cmd/show_source.rb:41]: Hmm, I'm not sure honestly.  Want me to ask on Mastodon and see what feedback I get?

Comment (st0012) [reply lib/irb/cmd/show_source.rb:41]: If we don't want to provide a different message for super definition, why do we need this `elsif` branch then 🤔 The message is exactly the same as the `else` condition's.

Comment (paulreece) [reply lib/irb/cmd/show_source.rb:41]: Lol, fair point!

Well what do you think?  Is it more helpful with the `elsif` branch adding `super` to the Error message or simply removing the `elsif` and having it return as it normally would when it hits the `else`?

Comment (st0012) [reply lib/irb/cmd/show_source.rb:41]: I think mentioning super is necessary because we do find the signatured method's source.

For example, in Pry:

- If it doesn't find the method's source, it prints `Error: Couldn't locate a definition for Foo#baz`
- But if it doesn't find the method's super method, it prints: `Error: No superclass found for #<UnboundMethod: Bar#baz() test.rb:6>`

I think we can tweak the message later, but IMO they have to be different.

Comment (paulreece) [reply lib/irb/cmd/show_source.rb:41]: Makes sense, changing now.

**Review by st0012 — APPROVED** [review]
I have a few ideas for small improvements and adopting the same flag to `edit`. But I think it's good to merge now. I'll let @tompng do another pass before pushing the button though.

@paulreece Thank you so much for the work on this PR, as well as all the feedback you provided during RubyConf ❤️

**Comment by paulreece** [comment]
> @paulreece Thank you so much for the work on this PR, as well as all the feedback you provided during RubyConf ❤️

@st0012 you're welcome!  It was a pleasure working with you and @tompng on this PR.  I'm happy to try to address other items from the RubyConf issues or other open bug issues if you'd like.  I really enjoy how constructive working in this repo is. :)

**Comment by st0012** [comment]
@paulreece It's not listed in the issues yet, but I'd really like to make `show_source` command provide more metadata about the method, like the ones from Pry:

```
Visibility: public
Signature: baz()
Number of lines: 3
```

You are definitely welcome to work on any other issues :-) But this one just came up to me today.

**Comment by paulreece** [comment]
>It's not listed in the issues yet, but I'd really like to make `show_source` command provide more metadata about the method, like the ones from Pry:
> 
> ```
> Visibility: public
> Signature: baz()
> Number of lines: 3
> ```
> 


Awesome!  I'll get started on this one!

File: lib/irb/source_finder.rb [hunk 11]
```diff
@@ -18,6 +18,9 @@ def initialize(irb_context)
 
     def find_source(signature)
       context_binding = @irb_context.workspace.binding
+      if signature.kind_of?(Array)
+        signature, s_count = signature
+      end
```
Comment (tompng) [inline lib/irb/source_finder.rb:23]: Can this be a separate method parameter?
```ruby
def find_source(signature, s_count = nil)
```

Comment (paulreece) [reply lib/irb/source_finder.rb:23]: Done!

File: lib/irb/source_finder.rb [hunk 12]
```diff
@@ -26,10 +26,18 @@ def find_source(signature)
...
+          super_owner = owner
+          s_count.times  do |s| 
+            if super_owner.superclass
+              super_owner = super_owner.superclass
+            end
+          end
+          owner = super_owner
         end
+        methods = owner.instance_methods + owner.private_instance_methods
+        file, line = owner.instance_method(method).source_location if methods.include?(method.to_sym)
```
Comment (st0012) [inline lib/irb/source_finder.rb:40]: Sorry that I only found this now: I think it's also possible to get the method through `super_method`? Something like:

```rb
target_method = owner.instance_method(method)

s_count.times do |s|
  target_method = target_method.super_method if target_method
end

file, line = target_method.source_location if target_method
```

File: lib/irb/source_finder.rb [hunk 13]
```diff
@@ -26,10 +26,18 @@ def find_source(signature)
...
       when /\A((?<receiver>.+)(\.|::))?(?<method>[^ :.]+)\z/ # method, receiver.method, receiver::method
         receiver = eval(Regexp.last_match[:receiver] || 'self', context_binding)
```
Comment (st0012) [inline lib/irb/source_finder.rb:42]: If we can use the above approach, we can support these signatures too.

Comment (paulreece) [reply lib/irb/source_finder.rb:42]: Just so I'm understanding this correctly.  We would be combining the two different case statements here when I implement the above approach? Or would I be adding this `target_method` logic into each one?

```ruby
 when /\A(?<owner>[A-Z]\w*(::[A-Z]\w*)*)#(?<method>[^ :.]+)\z/ # Class#method

 when /\A((?<receiver>.+)(\.|::))?(?<method>[^ :.]+)\z/ 
```

Comment (st0012) [reply lib/irb/source_finder.rb:42]: Not combining the case statements. I imagine you can extract the super lookup logic into a private method and use it in both cases instead.

Comment (paulreece) [reply lib/irb/source_finder.rb:42]: Got it

File: test/irb/test_cmd.rb [hunk 14]
```diff
@@ -469,6 +469,116 @@ def test_show_source_method
...
+            super
+          end
+        end
+      RUBY
+      File.write("#{@tmpdir}/bazbar.rb", code)
+      out, err = execute_lines(
+        "irb_load '#{@tmpdir}/bazbar.rb'\n",
+        "show_source Bar#foo -s",
+      )
+      assert_match(%r[Error: Couldn't locate a super definition for Bar#foo\n], out)
```
Comment (tompng) [inline test/irb/test_cmd.rb:536]: CI (ruby-core setup) is failing in this line.
You can reproduce it by removing `def test_show_source_method_multiple_s` and run `rake test TEST=test/irb/test_cmd.rb` in local.

The problem is that after other test, `Object::Baz` still remains with method `foo` defined.
In this case, `class Baz; end` is not a class definition but an open-class.
The test accidentally passed in other CI because source_location file `bazbarbob.rb` was already deleted.

I think the easiest way to fix this is to add these in all added tests
```ruby
ensure
  Object.remove_const :Bar
  Object.remove_const :Baz
  Object.remove_const :Bob
end
```

File: lib/irb/source_finder.rb (hunk 13 above)
Comment (paulreece) [reply lib/irb/source_finder.rb:32]: Implemented this and refactored `find_source` a bit as well.  Lmk what you think!

File: test/irb/test_cmd.rb (hunk 14 above)
Comment (paulreece) [reply test/irb/test_cmd.rb]: I ended up just changing the method names in each test since it wouldn't let me access the `remove_const` method now these pass even when commenting out `def test_show_source_method_multiple_s`

TruffleRuby CI is failing because of a gem I believe:
```ruby
In Gemfile:
  debug was resolved to 1.8.0, which depends on
    irb was resolved to 1.9.1, which depends on
      rdoc was resolved to 6.6.0, which depends on
        psych was resolved to 5.1.1.1, which depends on
          stringio
Error: The process '/home/runner/.rubies/truffleruby-head/bin/bundle' failed with exit code 5
```

Comment (st0012) [reply test/irb/test_cmd.rb]: (Don't worry about truffleruby for now, we're aware of the problem and have contacted related teams)

**Review by tompng — APPROVED** [review]

**Review by st0012 — APPROVED** [review]
Thank you so much!

**Comment by jemmaissroff** [comment]
👋 Looks like this PR has introduced test failures in the ruby/ruby master build (https://github.com/ruby/ruby/commits/master). Flagging in case you weren't aware

**Comment by st0012** [comment]
Thanks for flagging. I've just addressed the issues (hopefully) in https://github.com/ruby/irb/pull/793. But I think the commit was not synced to ruby/ruby?

**Comment by jemmaissroff** [comment]
Hmm interesting, it does seem it wasn't synced. Oddly too, it doesn't appear in the #alerts-sync channel in RubyLang slack which usually indicates sync failures. Perhaps we can try syncing it manually?

**Comment by paulreece** [comment]
Sorry about that y'all.  @st0012 Do you have any suggestions for the next PR I open in order to avoid the bug issues that the tests caused on this one?  Maybe I setup the tests a bit differently?

**Comment by st0012** [comment]
Just add new tests to the new file added in #793 should be enough 😄
